README.md
[![Build Status](https://travis-ci.org/n3okill/enfscompare.svg)](https://travis-ci.org/n3okill/enfscompare)
[![Build status](https://ci.appveyor.com/api/projects/status/hbrhk12g7gmse48t?svg=true)](https://ci.appveyor.com/project/n3okill/enfscompare)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/459632289dfe4dcfab15a69f625ec291)](https://www.codacy.com/app/n3okill/enfscompare)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfscompare.png)](https://nodei.co/npm/enfscompare/)

enfscompare
=========
Module that add compare functionality to to node fs module
 
This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)

Description
-----------
This module will add a method that allows the comparison of files and
folders.

- This module will add following methods to node fs module:
  * files
  * filesHash
  * filesSync
  * filesHashSync
  * dir
  * dirHash
  * dirSync
  * dirHashSync

Usage
-----
`enfscompare`

```js
    var enfscompare = require("enfscompare");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [files](#files)
- [filesSync](#filessync)
- [filesHash](#filesHash)
- [filesHashSync](#filesHashSync)
- [dir](#dir)
- [dirSync](#dirSync)
- [dirHash](#dirHash)
- [dirHashSync](#dirHashSync)


### files
### filesSync
  - **files(path1, path2, [options], callback)**
  - **filesSync(path1, path2, [options])**

> Compare files in a byte-to-byte form

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * chunkSize (Number) - the size in bytes used to read files and verify equality default to 
  (async: ReadStream._readableState.highWaterMark, sync: 8192)
  
```js
    enfscompare.files("/home/name/file.txt", "/home/name/file.txt", function(err, result){
        if(result===true){
            console.log("File is equal");
        }
    });
```


### dir
### dirSync
  - **dir(path1, path2, [options], callback)**
  - **dirSync(path1, path2, [options])**

> Compare files in a byte-to-byte form

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * chunkSize (Number) - the size in bytes used to read files and verify equality default to 
  (async: ReadStream._readableState.highWaterMark, sync: 8192)

```js
    enfscompare.dir("/home/name", "/home/another_folder_name", function(err, result){
        if(result===true){
            console.log("Directories are equal");
        }
    });
```


### filesHash
### filesHashSync
  - **filesHash(path1, path2, [options], callback)**
  - **filesHashSync(path1, path2, [options])**

> Compare files with an hash
NOTE: This method can't correctly tell you that the files are equal, but can tell you if they are different

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * hash (String) - the type of hash to use in comparison, default to 'sha512'
  * encoding (String) - the type of hash encoding used to compare the files, default to 'hex'
  * chunkSize (String) - the size in bytes used to read files default 8192 (Only in sync method)

```js
    var result = enfscompare.filesHashSync("/home/name/file.txt","/home/name/file.txt");
    if(result===true){
        console.log("File is equal");
    }
```



### dirHash
### dirHashSync
  - **dirHash(path1, path2, [options], callback)**
  - **dirHashSync(path1, path2, [options])**

> Compare files with an hash
NOTE: This method can't correctly tell you that the files are equal, but can tell you if they are different

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfspatch](https://www.npmjs.com/package/enfspatch))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * hash (String) - the type of hash to use in comparison, default to 'sha512'
  * encoding (String) - the type of hash encoding used to compare the files, default to 'hex'
  * chunkSize (String) - the size in bytes used to read files default 8192 (Only in sync method)

```js
    var result = enfscompare.filesHashSync("/home/name","/home/another_folder_name");
    if(result===true){
        console.log("Folders are equal");
    }
```


Benchmarks
----------
Run the files in benchmarks folder to check which algorithm is fastest in your machine with different file sizes


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-02-20
==================
  * code improvements
  * dependencies upgrade
  * major version
  
0.0.1 / 2016-04-12
==================

  * first release


test/dir.js
/**
 * @project enfscompare
 * @filename file.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016  Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 08-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description tests for file comparison
 */

/* global describe, it*/

"use strict";

const nodePath = require("path");
const comparator = require("../");


describe("enfscompare", function () {
    describe("> dir", function () {
        describe("> async", function () {
            describe("> hash", function () {
                it("should compare two equal directories", function (done) {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "lib");
                    comparator.dirHash(path1, path2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(true);
                        done();
                    });
                });
                it("should fail to compare two different directories", function (done) {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "test");
                    comparator.dirHash(path1, path2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(false);
                        done();
                    });
                });
            });
            describe("> byte to byte", function () {
                it("should compare two equal directories", function (done) {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "lib");
                    comparator.dir(path1, path2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(true);
                        done();
                    });
                });
                it("should fail to compare two different directories", function (done) {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "test");
                    comparator.dir(path1, path2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(false);
                        done();
                    });
                });
            });
        });
        describe("> sync", function () {
            describe("> hash", function () {
                it("should compare two equal directories", function () {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "lib");
                    let result = comparator.dirHashSync(path1, path2);
                    result.should.be.equal(true);
                });
                it("should fail to compare two different directories", function () {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "test");
                    let result = comparator.dirHashSync(path1, path2);
                    result.should.be.equal(false);
                });
            });
            describe("> byte to byte", function () {
                it("should compare two equal directories", function () {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "lib");
                    let result = comparator.dirSync(path1, path2);
                    result.should.be.equal(true);
                });
                it("should fail to compare two different directories", function () {
                    const path1 = nodePath.join(__dirname, "..", "lib");
                    const path2 = nodePath.join(__dirname, "..", "test");
                    let result = comparator.dirSync(path1, path2);
                    result.should.be.equal(false);
                });
            });
        });
    });
});


test/file.js
/**
 * @project enfscompare
 * @filename file.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016  Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 08-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description tests for file comparison
 */

/* global describe, it, before, after*/

"use strict";

const nodeCrypto = require("crypto");
const nodeOs = require("os");
const nodePath = require("path");
const enfs = require("enfspatch");
const comparator = require("../");
const rimraf = require("rimraf");

const MbSize = 10;


function prepareFiles(tmpPath, file, fileDiff, done) {
    let end1, end2;
    end1 = false;
    end2 = false;

    function ended() {
        if (end1 === true && end2 === true) {
            enfs.stat(file, done);
        }
    }

    function createFiles() {
        const stream1 = enfs.createWriteStream(file);
        const stream2 = enfs.createWriteStream(fileDiff);

        stream1.on("finish", () => {
            end1 = true;
            ended();
        });
        stream2.on("finish", () => {
            end2 = true;
            ended();
        });

        stream1.on("error", (err) => {
            stream2.destroy();
            done(err);
        });
        stream2.on("error", (err) => {
            stream1.destroy();
            done(err);
        });

        const size = 1024 * 1024; //Mb
        const steps = MbSize; // 10Mb
        const random1 = nodeCrypto.randomBytes(size);
        const random2 = nodeCrypto.randomBytes(size);

        let i = steps;
        while (i-- > 1) {
            stream1.write(random1);
            stream2.write(random1);
        }
        stream1.write(random1);
        stream2.write(random2);

        stream1.end();
        stream2.end();

    }

    enfs.mkdir(tmpPath, (err) => {
        if (err) {
            if (err.code === "EEXIST") {
                return rimraf(tmpPath + nodePath.sep + "*", (err) => {
                    if (err) {
                        return done(err);
                    }
                    createFiles();
                });
            }
            return done(err);
        }
        createFiles();
    });
}



describe("enfscompare", function () {
    describe("> file", function () {
        const tmpPath = nodePath.join(nodeOs.tmpDir(), "async_file");
        const file = nodePath.join(tmpPath, "file");
        const fileDiff = nodePath.join(tmpPath, "fileDiff");
        before(function (done) {
            //this.timeout = 5000;
            prepareFiles(tmpPath, file, fileDiff, done);
        });
        after(function (done) {
            rimraf(tmpPath, done);
        });
        describe("> async", function () {
            describe("> hash", function () {
                it("should compare two equal files", function (done) {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "index.js");
                    comparator.filesHash(file1, file2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(true);
                        done();
                    });
                });
                it("should fail to compare two different files", function (done) {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "package.json");
                    comparator.filesHash(file1, file2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(false);
                        done();
                    });
                });
                describe("> big files", function () {
                    it("should compare two equal files", function (done) {
                        comparator.filesHash(file, file, (err, result) => {
                            (err === null).should.be.equal(true);
                            result.should.be.equal(true);
                            done();
                        });
                    });
                    it("should fail to compare two different files", function (done) {
                        comparator.filesHash(file, fileDiff, (err, result) => {
                            (err === null).should.be.equal(true);
                            result.should.be.equal(false);
                            done();
                        });
                    });
                });
            });
            describe("> byte to byte", function () {
                it("should compare two equal files", function (done) {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "index.js");
                    comparator.files(file1, file2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(true);
                        done();
                    });
                });
                it("should fail to compare two different files", function (done) {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "package.json");
                    comparator.files(file1, file2, (err, result) => {
                        (err === null).should.be.equal(true);
                        result.should.be.equal(false);
                        done();
                    });
                });
                describe("> big files", function () {
                    it("should compare two equal files", function (done) {
                        comparator.files(file, file, (err, result) => {
                            (err === null).should.be.equal(true);
                            result.should.be.equal(true);
                            done();
                        });
                    });
                    it("should fail to compare two different files", function (done) {
                        comparator.files(file, fileDiff, (err, result) => {
                            (err === null).should.be.equal(true);
                            result.should.be.equal(false);
                            done();
                        });
                    });
                });
            });
        });
        describe("> sync", function () {
            describe("> hash", function () {
                it("should compare two equal files", function () {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "index.js");
                    let result = comparator.filesHashSync(file1, file2);
                    result.should.be.equal(true);
                });
                it("should fail to compare two different files", function () {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "package.json");
                    let result = comparator.filesHashSync(file1, file2);
                    result.should.be.equal(false);
                });
                describe("> big files", function () {
                    it("should compare two equal files", function () {
                        let result = comparator.filesHashSync(file, file);
                        result.should.be.equal(true);
                    });
                    it("should fail to compare two different files", function () {
                        let result = comparator.filesHashSync(file, fileDiff);
                        result.should.be.equal(false);
                    });
                });
            });
            describe("> byte to byte", function () {
                it("should compare two equal files", function () {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "index.js");
                    let result = comparator.filesSync(file1, file2);
                    result.should.be.equal(true);
                });
                it("should fail to compare two different files", function () {
                    const file1 = nodePath.join(__dirname, "..", "index.js");
                    const file2 = nodePath.join(__dirname, "..", "package.json");
                    let result = comparator.filesSync(file1, file2);
                    result.should.be.equal(false);
                });
                describe("> big files", function () {
                    it("should compare two equal files", function () {
                        let result = comparator.filesSync(file, file);
                        result.should.be.equal(true);
                    });
                    it("should fail to compare two different files", function () {
                        let result = comparator.filesSync(file, fileDiff);
                        result.should.be.equal(false);
                    });
                });
            });
        });
    });
});



lib/sync/dir.js
/**
 * @project enfscompare
 * @filename dir.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 07-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description sync directory comparison
 */


"use strict";


const nodePath = require("path");
const enfs = require("enfspatch");
const enfsList = require("enfslist");
const compareFile = require("./file");



function statCompare(dir1, dir2, options) {
    options.stat = options.dereference ? options.fs.statSync : options.fs.lstatSync;

    const stat1 = options.stat(dir1);
    const stat2 = options.stat(dir2);
    if (!stat1.isDirectory()) {
        throw new Error("path1 is not a directory.");
    }
    if (!stat2.isDirectory()) {
        throw new Error("path2 is not a directory.");
    }
    return true;
}


function compareFiles(type, dir1, dir2, options) {
    options.list = options.fs.listSync || enfsList.listSync;

    const items1 = options.list(dir1, options);
    const items2 = options.list(dir2, options);
    if (items1.length !== items2.length) {
        return false;
    }
    const files1 = items1.filter((f) => f.stat.isFile());
    const files2 = items2.filter((f) => f.stat.isFile());
    const files2Reduced = files2.map((i) => i.path.replace(dir2, ""));
    let size = files1.length;
    while (size-- > 0) {
        const pathReduced = files1[size].path.replace(dir1, "");
        const index = files2Reduced.indexOf(pathReduced);
        if (index === -1) {
            return false;
        }
        let result = compareFile[type](files1[size].path, files2[index].path, options);
        if (result === false) {
            return false;
        }
    }
    return true;
}


/**
 * Compare two directories in a byte-to-byte file comparison
 * Note: This method will compare all sub-folder's
 * @param {string} path1 - the path to the first file
 * @param {string} path2 - the path to the second file
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used will use enfs by default
 *              {Number} opt.chunkSize - the size in bytes used to read files and verify equality default to 8192
 * @return {Error|Array}
 */
function byte(path1, path2, opt) {
    let options;
    options = opt || {};
    options.fs = options.fs || enfs;
    options.dereference = options.dereference === true;

    const dir1 = nodePath.resolve(path1);
    const dir2 = nodePath.resolve(path2);
    if (statCompare(dir1, dir2, options) === true) {
        return compareFiles("byte", dir1, dir2, options);
    }
}

/**
 * Compare two directories with a hash file comparison
 * Note: This method will compare all sub-folder's
 * @param {string} path1 - the path to the first file
 * @param {string} path2 - the path to the second file
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used will use enfs by default
 *              {string} opt.hash - the type of hash to use in comparison, default to 'sha512'
 *              {string} opt.encoding - the type of hash encoding used to compare the files, default to 'hex'
 *              {Number} opt.chunkSize - the size in bytes used to read files default 8192
 * @return {Error|Array}
 */
function hash(path1, path2, opt) {
    let options;
    options = opt || {};
    options.fs = options.fs || enfs;
    options.dereference = options.dereference === true;

    const dir1 = nodePath.resolve(path1);
    const dir2 = nodePath.resolve(path2);
    if (statCompare(dir1, dir2, options) === true) {
        return compareFiles("hash", dir1, dir2, options);
    }
}



module.exports = {
    byte,
    hash
};


lib/sync/file.js
/**
 * @project enfscompare
 * @filename file.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 07-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description sync file comparison
 */

"use strict";

const nodePath = require("path");
const nodeCrypto = require("crypto");
const enfs = require("enfspatch");



function statCompare(file1, file2, options) {
    const stat1 = options.stat(file1);
    const stat2 = options.stat(file2);
    if (!stat1.isFile()) {
        throw new Error("file1 is not a file.");
    }
    if (!stat2.isFile()) {
        throw new Error("file2 is not a file.");
    }
    return stat1.size === stat2.size;
}

function compareFiles(file1, file2, options) {
    const fd1 = options.fs.openSync(file1, "r");
    const fd2 = options.fs.openSync(file2, "r");
    const chunkSize = options.chunkSize && !isNaN(options.chunkSize) ? options.chunkSize : 8192;
    try {
        let done1, done2, data1, data2;
        done1 = false;
        done2 = false;
        data1 = new Buffer(0);
        data2 = new Buffer(0);
        while (!done1 || !done2) {
            let buf1 = new Buffer(chunkSize);
            let buf2 = new Buffer(chunkSize);
            buf1.fill(0);
            buf2.fill(0);
            let bytesRead1 = options.fs.readSync(fd1, buf1, 0, chunkSize);
            if (bytesRead1) {
                data1 = Buffer.concat([data1, buf1]);
            }
            let bytesRead2 = options.fs.readSync(fd2, buf2, 0, chunkSize);
            if (bytesRead2) {
                data2 = Buffer.concat([data2, buf2]);
            }
            done1 = bytesRead1 === 0;
            done2 = bytesRead2 === 0;
            if (data1.length >= chunkSize && data2.length >= chunkSize) {
                let b1 = data1.slice(0, chunkSize);
                let b2 = data2.slice(0, chunkSize);
                if (!b1.equals(b2)) {
                    return false;
                }
                data1 = data1.slice(chunkSize);
                data2 = data2.slice(chunkSize);
            }
        }
        return true;
    } catch (err) {
        throw err;
    } finally {
        options.fs.closeSync(fd1);
        options.fs.closeSync(fd2);
    }
}

/**
 * Compare two files in a byte-to-byte comparison
 * @param {string} path1 - the path to the first file
 * @param {string} path2 - the path to the second file
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used will use enfs by default
 *              {Number} opt.chunkSize - the size in bytes used to read files and verify equality default 8192
 * @return {Error|Array}
 */
function byte(path1, path2, opt) {
    let options;
    const file1 = nodePath.resolve(path1);
    const file2 = nodePath.resolve(path2);

    options = opt || {};
    options.fs = options.fs || enfs;
    options.dereference = options.dereference === true;
    options.stat = options.dereference ? options.fs.statSync : options.fs.lstatSync;

    if (!statCompare(file1, file2, options)) {
        return false;
    }
    return compareFiles(file1, file2, options);
}


/**
 * Compare two files in a hash comparison
 * @param {string} path1 - the path to the first file
 * @param {string} path2 - the path to the second file
 * @param {object} opt - various options for list module
 *              {bool} opt.dereference - defines the type of stat that will be used in files
 *              {object} opt.fs - the fs module to be used will use enfs by default
 *              {string} opt.hash - the type of hash to use in comparison, default to 'sha512'
 *              {string} opt.encoding - the type of hash encoding used to compare the files, default to 'hex'
 *              {Number} opt.chunkSize - the size in bytes used to read files default 8192
 * @return {Error|Array}
 */
function hash(path1, path2, opt) {
    let options;
    const file1 = nodePath.resolve(path1);
    const file2 = nodePath.resolve(path2);

    options = opt || {};
    options.fs = options.fs || enfs;
    options.dereference = options.dereference === true;
    options.stat = options.dereference ? options.fs.statSync : options.fs.lstatSync;
    options.hash = options.hash || "sha512";
    options.encoding = options.encoding || "hex";
    const chunkSize = options.chunkSize && !isNaN(options.chunkSize) ? options.chunkSize : 8192;

    if (!statCompare(file1, file2, options)) {
        return false;
    }

    const hash1 = nodeCrypto.createHash(options.hash);
    const hash2 = nodeCrypto.createHash(options.hash);
    hash1.setEncoding(options.encoding);
    hash2.setEncoding(options.encoding);
    const fd1 = options.fs.openSync(file1, "r");
    const fd2 = options.fs.openSync(file2, "r");

    try {
        let done1, done2;
        done1 = false;
        done2 = false;
        while (!done1 || !done2) {
            let buf1 = new Buffer(chunkSize);
            let buf2 = new Buffer(chunkSize);
            buf1.fill(0);
            buf2.fill(0);
            let bytesRead1 = options.fs.readSync(fd1, buf1, 0, chunkSize);
            if (bytesRead1) {
                hash1.update(buf1);
            }
            let bytesRead2 = options.fs.readSync(fd2, buf2, 0, chunkSize);
            if (bytesRead2) {
                hash2.update(buf2);
            }
            done1 = bytesRead1 === 0;
            done2 = bytesRead2 === 0;
        }
        hash1.end();
        hash2.end();
        return hash1.read() === hash2.read();
    } catch (err) {
        throw err;
    } finally {
        options.fs.closeSync(fd1);
        options.fs.closeSync(fd2);
    }
}

module.exports = {
    byte,
    hash
};




lib/enfscompare.js
/**
 * @project enfscompare
 * @filename enfscompare.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 07-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description enfscompare entry point
 */

"use strict";

const async = {
    files: require("./async/file"),
    dir: require("./async/dir")
};
const sync = {
    files: require("./sync/file"),
    dir: require("./sync/dir")
};

module.exports = {
    files: async.files.byte,
    filesHash: async.files.hash,
    filesSync: sync.files.byte,
    filesHashSync: sync.files.hash,
    dir: async.dir.byte,
    dirHash: async.dir.hash,
    dirSync: sync.dir.byte,
    dirHashSync: sync.dir.hash
};


benchmarks/benchmarks_async.js
/**
 * @project enfscompare
 * @filename benchmarks_async.js
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016  Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 08-04-2016.
 * @version 0.0.1
 * @since 0.0.1
 * @description display the different times taken to compare files of different sizes with the 2 methods available
 * (hash and byte-to-byte)
 * Run this in your machine to see what is fastest, normally byte-to-byte in different files will be the fastest
 */

"use strict";

const nodePath = require("path");
const nodeOs = require("os");
const nodeCrypto = require("crypto");
const rimraf = require("rimraf");
const enfs = require("enfspatch");
const comparator = require("../");

const tmpPath = nodePath.join(nodeOs.tmpdir(), "enfscomparebenchmarks");


const FILES = {
    SMALL: {
        FILE: nodePath.join(tmpPath, "small_file"),
        DIFF: nodePath.join(tmpPath, "small_diff")
    },
    MEDIUM: {
        FILE: nodePath.join(tmpPath, "medium_file"),
        DIFF: nodePath.join(tmpPath, "medium_diff")
    },
    BIG: {
        FILE: nodePath.join(tmpPath, "big_file"),
        DIFF: nodePath.join(tmpPath, "big_diff")
    }
};
const SIZE = 1024 * 1024; //Mb
const MBSIZE = {
    SMALL: 1,       //1Mb
    MEDIUM: 100,    //100Mb
    BIG: 1000       //1Gb
};

const STATS = {
    SMALL: {
        SIZE: 0,
        HASH: [],
        BYTE: []
    },
    MEDIUM: {
        SIZE: 0,
        HASH: [],
        BYTE: []
    },
    BIG: {
        SIZE: 0,
        HASH: [],
        BYTE: []
    }
};

function onError(err) {
    if (err) {
        rimraf.sync(tmpPath);
        throw err;
    }
}


function createFile(name, type, random1, random2, done) {
    const stream1 = enfs.createWriteStream(FILES[type][name]);

    stream1.on("finish", () => {
        enfs.stat(FILES[type][name], (err, stat) => {
            onError(err);
            STATS[type].SIZE = stat.size;
            done();
        });
    });

    stream1.on("error", (err) => {
        onError(err);
        done(err);
    });

    let i = MBSIZE[type];
    while (i-- > 1) {
        stream1.write(random1);
    }
    stream1.write(random2);

    stream1.end();
}


function createType(type, done) {
    const random1 = nodeCrypto.randomBytes(SIZE);
    const random2 = nodeCrypto.randomBytes(SIZE);
    createFile("FILE", type, random1, random1, () => {
        createFile("DIFF", type, random1, random2, () => {
            done();
        });
    });
}

function createFiles(done) {
    createType("SMALL", () => {
        createType("MEDIUM", () => {
            createType("BIG", () => {
                done();
            });
        });
    });
}

function createDir(done) {
    enfs.mkdir(tmpPath, (err) => {
        if (err) {
            if (err.code === "EEXIST") {
                return createFiles(done);
            }
            onError(err);
        }
        createFiles(done);
    });
}


function runComparator(fn, type, statType, done) {
    let start = Date.now();
    comparator[fn](FILES[type].FILE, FILES[type].FILE, (err, result) => {
        onError(err);
        if (result !== true) {
            throw new Error(type + " - " + fn + " - returned invalid result.");
        }
        STATS[type][statType].push(start, Date.now());
        start = Date.now();
        comparator[fn](FILES[type].FILE, FILES[type].DIFF, (err2, result2) => {
            if (result2 !== false) {
                throw new Error(type + " - " + fn + " - returned invalid result.");
            }
            STATS[type][statType].push(start, Date.now());
            onError(err2);
            done();
        });
    });
}

function runType(type, done) {
    runComparator("filesHash", type, "HASH", () => {
        runComparator("files", type, "BYTE", () => {
            done();
        });
    });
}


function benchmarks(done) {
    console.log("Running small");
    runType("SMALL", () => {
        console.log("Running medium");
        runType("MEDIUM", () => {
            console.log("Running big");
            runType("BIG", () => {
                done();
            });
        });
    });
}


function formatTime(millis) {
    let date = new Date(millis);
    return date.getHours() + ":" + date.getMinutes() + ":" + date.getSeconds() + "." + date.getMilliseconds();
}


function showComparator(type, statType, done) {
    let equal, diff;
    equal = STATS[type][statType][1] - STATS[type][statType][0];
    diff = STATS[type][statType][3] - STATS[type][statType][2];
    console.log(type + " - " + statType + " - equal: " + formatTime(equal));
    console.log(type + " - " + statType + " - diff: " + formatTime(diff));
    done();
}


function showType(type, done) {
    showComparator(type, "HASH", () => {
        showComparator(type, "BYTE", () => {
            done();
        });
    });
}

function runShowStats(done) {
    showType("SMALL", () => {
        showType("MEDIUM", () => {
            showType("BIG", () => {
                done();
            });
        });
    });
}

function showStats(done) {
    runShowStats(() => {
        console.log("Benchmark ended.");
        done();
    });
}


setImmediate(() => {
    const startTime = Date.now();
    console.log("Starting files creation.");
    console.time("File Creation");
    createDir(() => {
        console.timeEnd("File Creation");
        //console.log("Files created.");
        console.log("Starting to run benchmarks.");
        console.time("Benchmarks");
        benchmarks(() => {
            console.timeEnd("Benchmarks");
            console.log("Total time: " + formatTime(Date.now() - startTime));
            showStats(() => {
                rimraf.sync(tmpPath);
            });
        });
    });
});
